// @ngrx/effects/esm2015/src/lifecycle_hooks.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * \@description
 * Interface to set an identifier for effect instances.
 *
 * By default, each Effects class is registered
 * once regardless of how many times the Effect class
 * is loaded. By implementing this interface, you define
 * a unique identifier to register an Effects class instance
 * multiple times.
 *
 * \@usageNotes
 *
 * ### Set an identifier for an Effects class
 *
 * ```ts
 * class EffectWithIdentifier implements OnIdentifyEffects {
 *  constructor(private effectIdentifier: string) {}
 *
 *  ngrxOnIdentifyEffects() {
 *    return this.effectIdentifier;
 *  }
 *
 * ```
 * @record
 */
export function OnIdentifyEffects() { }
if (false) {
    /**
     * \@description
     * String identifier to differentiate effect instances.
     * @return {?}
     */
    OnIdentifyEffects.prototype.ngrxOnIdentifyEffects = function () { };
}
/** @type {?} */
export const onIdentifyEffectsKey = 'ngrxOnIdentifyEffects';
/**
 * \@description
 * Interface to control the lifecycle of effects.
 *
 * By default, effects are merged and subscribed to the store. Implement the OnRunEffects interface to control the lifecycle of the resolved effects.
 *
 * \@usageNotes
 *
 * ### Implement the OnRunEffects interface on an Effects class
 *
 * ```ts
 * export class UserEffects implements OnRunEffects {
 *   constructor(private actions$: Actions) {}
 *
 *   ngrxOnRunEffects(resolvedEffects$: Observable<EffectNotification>) {
 *     return this.actions$.pipe(
 *       ofType('LOGGED_IN'),
 *       exhaustMap(() =>
 *         resolvedEffects$.pipe(
 *           takeUntil(this.actions$.pipe(ofType('LOGGED_OUT')))
 *         )
 *       )
 *     );
 *   }
 * }
 * ```
 * @record
 */
export function OnRunEffects() { }
if (false) {
    /**
     * \@description
     * Method to control the lifecycle of effects.
     * @param {?} resolvedEffects$
     * @return {?}
     */
    OnRunEffects.prototype.ngrxOnRunEffects = function (resolvedEffects$) { };
}
/** @type {?} */
export const onRunEffectsKey = 'ngrxOnRunEffects';
/**
 * \@description
 * Interface to dispatch an action after effect registration.
 *
 * Implement this interface to dispatch a custom action after
 * the effect has been added. You can listen to this action
 * in the rest of the application to execute something after
 * the effect is registered.
 *
 * \@usageNotes
 *
 * ### Set an identifier for an Effects class
 *
 * ```ts
 * class EffectWithInitAction implements OnInitEffects {
 *  ngrxOnInitEffects() {
 *    return { type: '[EffectWithInitAction] Init' };
 *  }
 * ```
 * @record
 */
export function OnInitEffects() { }
if (false) {
    /**
     * \@description
     * Action to be dispatched after the effect is registered.
     * @return {?}
     */
    OnInitEffects.prototype.ngrxOnInitEffects = function () { };
}
/** @type {?} */
export const onInitEffects = 'ngrxOnInitEffects';
